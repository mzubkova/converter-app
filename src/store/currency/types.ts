import { ActionType } from "typesafe-actions";
import * as Actions from "./actions";

export type IExchangeRate = {
  ccy: string;
  base_ccy: string;
  buy: string;
  sale: string;
};

export type ICurrencyStore = {
  exchange_rates: Array<IExchangeRate>;
  isLoading: boolean;
};

type TActions = typeof Actions;

export type TCurrencyActions = ActionType<TActions>;
