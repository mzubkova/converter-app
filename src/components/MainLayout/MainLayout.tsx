import React from "react";
import { Typography } from "@mui/material";
import CurrencyConverter from "../CurrencyConverter";
import { StMainLayout } from "../styled/common";

const MainLayout = () => (
  <StMainLayout maxWidth="xl" height="100vh" padding="30px 0">
    <Typography
      component="h1"
      color="#041c32"
      fontSize="50px"
      textAlign="center"
      data-testid="header-title"
      margin="0 0 30px"
    >
      Converter app
    </Typography>
    <CurrencyConverter />
  </StMainLayout>
);

export default MainLayout;
