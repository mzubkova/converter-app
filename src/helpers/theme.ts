import { createTheme } from "@mui/material/styles";

export const theme = createTheme({
  palette: {
    primary: {
      main: "#0971f1",
      contrastText: "#053e85",
    },
    darkGrey: {
      main: "#0f0e0e",
      contrastText: "#053e85",
    },
    pink: {
      main: "#ff5677",
      contrastText: "#053e85",
    },
    blue: {
      main: "#22577a",
      contrastText: "#80ed99",
    },
  },
});
