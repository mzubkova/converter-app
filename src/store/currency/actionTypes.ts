export enum ActionTypes {
  GET_EXCHANGE_RATES = "@@currency/GET_EXCHANGE_RATES",
  GET_EXCHANGE_RATES_SUCCESS = "@@currency/GET_EXCHANGE_RATES_SUCCESS",
  GET_EXCHANGE_RATES_ERROR = "@@currency/GET_EXCHANGE_RATES_ERROR",
}
