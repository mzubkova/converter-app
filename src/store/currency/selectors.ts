import { createSelector } from "reselect";
import { ApplicationState } from "../types";
import * as Types from "./types";

export const getCurrencyState = (
  state: ApplicationState
): Types.ICurrencyStore => state.currency;
export const getExchangeRatesState = createSelector(
  getCurrencyState,
  (currency: Types.ICurrencyStore): Array<any> => currency.exchange_rates
);
export const getIsLoading = createSelector(
  getCurrencyState,
  (currency: Types.ICurrencyStore): boolean => currency.isLoading
);
