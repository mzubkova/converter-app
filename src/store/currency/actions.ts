import { action } from "typesafe-actions";
import { ActionTypes as AT } from "./actionTypes";
import { IExchangeRate } from "./types";

export const getExchangeRates = () => action(AT.GET_EXCHANGE_RATES);
export const getExchangeRatesSuccess = (payload: Array<IExchangeRate>) =>
  action(AT.GET_EXCHANGE_RATES_SUCCESS, payload);
export const getExchangeRatesError = () => action(AT.GET_EXCHANGE_RATES_ERROR);
