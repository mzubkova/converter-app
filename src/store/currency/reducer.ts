import { Reducer } from "redux";
import { ActionTypes as AT } from "./actionTypes";
import * as Types from "./types";

export const initialState: Types.ICurrencyStore = {
  exchange_rates: [],
  isLoading: false,
};

export const currencyReducer: Reducer<
  Types.ICurrencyStore,
  Types.TCurrencyActions
> = (state = initialState, action) => {
  switch (action.type) {
    case AT.GET_EXCHANGE_RATES:
      return {
        ...state,
        isLoading: true,
      };
    case AT.GET_EXCHANGE_RATES_SUCCESS:
      return {
        ...state,
        exchange_rates: action.payload,
        isLoading: false,
      };
    case AT.GET_EXCHANGE_RATES_ERROR:
      return {
        ...state,
        isLoading: false,
      };
    default:
      return state;
  }
};
