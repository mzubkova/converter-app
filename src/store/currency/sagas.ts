import { put, takeEvery, call } from "redux-saga/effects";
import { ActionTypes as AT } from "./actionTypes";
import { SagaIterator } from "@redux-saga/core";
import * as actions from "./actions";
import * as request from "../../api/index";

export function* getExchangeRates(): SagaIterator {
  try {
    const data = yield call(request.getRequest);
    yield put(actions.getExchangeRatesSuccess(data));
  } catch {
    yield put(actions.getExchangeRatesError());
  }
}

export default function* actionWatcher(): SagaIterator {
  yield takeEvery(AT.GET_EXCHANGE_RATES, getExchangeRates);
}
