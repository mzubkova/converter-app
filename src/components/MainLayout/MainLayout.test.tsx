import * as React from "react";
import MainLayout from "../MainLayout";
import { shallowRender, mockStore } from "../../../__tests__/helpers";
import { initialState } from "../../store/currency/reducer";
import { screen } from "@testing-library/react";

describe("MainLayout", () => {
  let initialStore = {};
  let component: any;
  beforeEach(() => {
    initialStore = mockStore({ currency: initialState });
    component = shallowRender(<MainLayout />, initialStore);
  });
  it("should render without crashing", () => {
    expect(component).toMatchSnapshot();
  });
  it("should mount text in button", () => {
    expect(screen.getByRole("button").textContent).toBe("Convert");
  });
  it("should mount h1", () => {
    expect(screen.getByTestId("header-title")).toBeInTheDocument();
  });
});
