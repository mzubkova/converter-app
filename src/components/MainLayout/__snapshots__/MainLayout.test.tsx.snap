// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`MainLayout should render without crashing 1`] = `
Object {
  "asFragment": [Function],
  "baseElement": <body>
    <div>
      <div
        class="MuiContainer-root MuiContainer-maxWidthXl common__StMainLayout-sc-1gfg3wy-0 btoDZQ css-19r6kue-MuiContainer-root"
        height="100vh"
        padding="30px 0"
      >
        <h1
          class="MuiTypography-root MuiTypography-body1 css-g5l3it-MuiTypography-root"
          data-testid="header-title"
        >
          Converter app
        </h1>
        <div
          class="styled__StCurrencyConverter-sc-11jaufk-0 kgKJWc"
        >
          <div
            class="common__StFlex-sc-1gfg3wy-3 cmwpvq"
          >
            <div
              class="MuiFormControl-root MuiTextField-root common__StInput-sc-1gfg3wy-2 jSJBEF css-1u3bzj6-MuiFormControl-root-MuiTextField-root"
            >
              <label
                class="MuiInputLabel-root MuiInputLabel-formControl MuiInputLabel-animated MuiInputLabel-outlined MuiFormLabel-root MuiFormLabel-colorPrimary css-5mltfn-MuiFormLabel-root-MuiInputLabel-root"
                data-shrink="false"
                for=":r0:"
                id=":r0:-label"
              >
                1 usd in uah
              </label>
              <div
                class="MuiOutlinedInput-root MuiInputBase-root MuiInputBase-colorPrimary MuiInputBase-formControl css-n8ea9h-MuiInputBase-root-MuiOutlinedInput-root"
              >
                <input
                  aria-invalid="false"
                  class="MuiOutlinedInput-input MuiInputBase-input css-1t8l2tu-MuiInputBase-input-MuiOutlinedInput-input"
                  id=":r0:"
                  type="text"
                  value=""
                />
                <fieldset
                  aria-hidden="true"
                  class="MuiOutlinedInput-notchedOutline css-1d3z3hw-MuiOutlinedInput-notchedOutline"
                >
                  <legend
                    class="css-1ftyaf0"
                  >
                    <span>
                      1 usd in uah
                    </span>
                  </legend>
                </fieldset>
              </div>
            </div>
            <button
              class="MuiButton-root MuiButton-contained MuiButton-containedBlue MuiButton-sizeMedium MuiButton-containedSizeMedium MuiButtonBase-root common__StButton-sc-1gfg3wy-1 ceNtmp css-1c27a6x-MuiButtonBase-root-MuiButton-root"
              margin="20px 0 20px 20px"
              tabindex="0"
              type="button"
            >
              Convert
              <span
                class="MuiTouchRipple-root css-8je8zh-MuiTouchRipple-root"
              />
            </button>
          </div>
          <p
            class="MuiTypography-root MuiTypography-body1 css-lc51lz-MuiTypography-root"
          />
          <p
            class="MuiTypography-root MuiTypography-body1 css-1gvvk8k-MuiTypography-root"
          >
            Result: 
          </p>
        </div>
      </div>
    </div>
  </body>,
  "container": <div>
    <div
      class="MuiContainer-root MuiContainer-maxWidthXl common__StMainLayout-sc-1gfg3wy-0 btoDZQ css-19r6kue-MuiContainer-root"
      height="100vh"
      padding="30px 0"
    >
      <h1
        class="MuiTypography-root MuiTypography-body1 css-g5l3it-MuiTypography-root"
        data-testid="header-title"
      >
        Converter app
      </h1>
      <div
        class="styled__StCurrencyConverter-sc-11jaufk-0 kgKJWc"
      >
        <div
          class="common__StFlex-sc-1gfg3wy-3 cmwpvq"
        >
          <div
            class="MuiFormControl-root MuiTextField-root common__StInput-sc-1gfg3wy-2 jSJBEF css-1u3bzj6-MuiFormControl-root-MuiTextField-root"
          >
            <label
              class="MuiInputLabel-root MuiInputLabel-formControl MuiInputLabel-animated MuiInputLabel-outlined MuiFormLabel-root MuiFormLabel-colorPrimary css-5mltfn-MuiFormLabel-root-MuiInputLabel-root"
              data-shrink="false"
              for=":r0:"
              id=":r0:-label"
            >
              1 usd in uah
            </label>
            <div
              class="MuiOutlinedInput-root MuiInputBase-root MuiInputBase-colorPrimary MuiInputBase-formControl css-n8ea9h-MuiInputBase-root-MuiOutlinedInput-root"
            >
              <input
                aria-invalid="false"
                class="MuiOutlinedInput-input MuiInputBase-input css-1t8l2tu-MuiInputBase-input-MuiOutlinedInput-input"
                id=":r0:"
                type="text"
                value=""
              />
              <fieldset
                aria-hidden="true"
                class="MuiOutlinedInput-notchedOutline css-1d3z3hw-MuiOutlinedInput-notchedOutline"
              >
                <legend
                  class="css-1ftyaf0"
                >
                  <span>
                    1 usd in uah
                  </span>
                </legend>
              </fieldset>
            </div>
          </div>
          <button
            class="MuiButton-root MuiButton-contained MuiButton-containedBlue MuiButton-sizeMedium MuiButton-containedSizeMedium MuiButtonBase-root common__StButton-sc-1gfg3wy-1 ceNtmp css-1c27a6x-MuiButtonBase-root-MuiButton-root"
            margin="20px 0 20px 20px"
            tabindex="0"
            type="button"
          >
            Convert
            <span
              class="MuiTouchRipple-root css-8je8zh-MuiTouchRipple-root"
            />
          </button>
        </div>
        <p
          class="MuiTypography-root MuiTypography-body1 css-lc51lz-MuiTypography-root"
        />
        <p
          class="MuiTypography-root MuiTypography-body1 css-1gvvk8k-MuiTypography-root"
        >
          Result: 
        </p>
      </div>
    </div>
  </div>,
  "debug": [Function],
  "findAllByAltText": [Function],
  "findAllByDisplayValue": [Function],
  "findAllByLabelText": [Function],
  "findAllByPlaceholderText": [Function],
  "findAllByRole": [Function],
  "findAllByTestId": [Function],
  "findAllByText": [Function],
  "findAllByTitle": [Function],
  "findByAltText": [Function],
  "findByDisplayValue": [Function],
  "findByLabelText": [Function],
  "findByPlaceholderText": [Function],
  "findByRole": [Function],
  "findByTestId": [Function],
  "findByText": [Function],
  "findByTitle": [Function],
  "getAllByAltText": [Function],
  "getAllByDisplayValue": [Function],
  "getAllByLabelText": [Function],
  "getAllByPlaceholderText": [Function],
  "getAllByRole": [Function],
  "getAllByTestId": [Function],
  "getAllByText": [Function],
  "getAllByTitle": [Function],
  "getByAltText": [Function],
  "getByDisplayValue": [Function],
  "getByLabelText": [Function],
  "getByPlaceholderText": [Function],
  "getByRole": [Function],
  "getByTestId": [Function],
  "getByText": [Function],
  "getByTitle": [Function],
  "queryAllByAltText": [Function],
  "queryAllByDisplayValue": [Function],
  "queryAllByLabelText": [Function],
  "queryAllByPlaceholderText": [Function],
  "queryAllByRole": [Function],
  "queryAllByTestId": [Function],
  "queryAllByText": [Function],
  "queryAllByTitle": [Function],
  "queryByAltText": [Function],
  "queryByDisplayValue": [Function],
  "queryByLabelText": [Function],
  "queryByPlaceholderText": [Function],
  "queryByRole": [Function],
  "queryByTestId": [Function],
  "queryByText": [Function],
  "queryByTitle": [Function],
  "rerender": [Function],
  "unmount": [Function],
}
`;
